File: app/src/main/java/com/tutv/android/ui/series_carrousel/tv_banner/TvBannerPresenter.java
package com.tutv.android.ui.series_carrousel.tv_banner;

import com.tutv.android.domain.Series;
import com.tutv.android.repository.SeriesRepository;
import com.tutv.android.ui.series.SeriesView;
import com.tutv.android.ui.tv_poster_list.tv_poster.TvPosterView;
import com.tutv.android.utils.schedulers.BaseSchedulerProvider;

import java.lang.ref.WeakReference;

import io.reactivex.disposables.CompositeDisposable;

public class TvBannerPresenter {
    private final WeakReference<TvBannerView> view;
    private final SeriesRepository seriesRepository;
    private final BaseSchedulerProvider schedulerProvider;

    private final CompositeDisposable disposables;

    private Series series;

    public TvBannerPresenter(TvBannerView view, SeriesRepository seriesRepository, BaseSchedulerProvider schedulerProvider) {
        this.view = new WeakReference<>(view);
        this.seriesRepository = seriesRepository;
        this.schedulerProvider = schedulerProvider;
        disposables = new CompositeDisposable();
    }

    public void bind(Series series) {
        this.series = series;

        if(view.get() != null) {
            view.get().setImageUrl(series.getBannerUrl());
            view.get().setTitle(series.getName());
            view.get().setSubtitle(series.getFollowers() + " followers");

            view.get().setSeriesFollowed(series.getLoggedInUserFollows() != null && series.getLoggedInUserFollows());
        }
    }

    public void onClick() {
        if(view.get() != null)
            view.get().navigateToSeriesPage(series.getId());
    }

    public void onViewDetached() {
        disposables.dispose();
    }

    public void onSeriesFollowClicked() {
        if(series.getLoggedInUserFollows() == null || !series.getLoggedInUserFollows()) {
            disposables.add(seriesRepository.setFollowSeries(series)
                    .observeOn(schedulerProvider.ui())
                    .subscribe(this::onSeriesFollowed, this::onSeriesFollowedError));
        } else {
            disposables.add(seriesRepository.unfollowSeries(series)
                    .observeOn(schedulerProvider.ui())
                    .subscribe(this::onSeriesFollowed, this::onSeriesFollowedError));
        }
    }

    private void onSeriesFollowedError(Throwable throwable) {
        if(view.get() != null) {
            view.get().showError("Error al dejar seguir la serie :(");
        }
    }

    private void onSeriesFollowed(Series series) {
        if(view.get() != null) {
            view.get().setSubtitle(series.getFollowers() + " followers");
            view.get().setSeriesFollowed(series.getLoggedInUserFollows());
        }
    }
}

File: app/src/main/java/com/tutv/android/ui/series_carrousel/tv_banner/TvBannerPresenter.kt
package com.tutv.android.ui.series_carrousel.tv_banner

import com.tutv.android.domain.Series
import com.tutv.android.repository.SeriesRepository
import com.tutv.android.utils.schedulers.BaseSchedulerProvider
import io.reactivex.disposables.CompositeDisposable
import java.lang.ref.WeakReference

class TvBannerPresenter(
        view: TvBannerView,
        var seriesRepository: SeriesRepository,
        var schedulerProvider: BaseSchedulerProvider) {

    private val view: WeakReference<TvBannerView?> = WeakReference(view)
    private val disposables: CompositeDisposable = CompositeDisposable()
    private var series: Series? = null

    fun bind(series: Series?) {
        this.series = series
        view.get()?.setImageUrl(series?.bannerUrl)
        view.get()?.setTitle(series?.name)
        view.get()?.setSubtitle(series?.followers.toString() + if (series?.followers == 1) " follower" else " followers")
        view.get()?.setSeriesFollowed(series?.loggedInUserFollows)
    }

    fun onClick() {
        view.get()?.navigateToSeriesPage(series?.id)
    }

    fun onViewDetached() {
        disposables.dispose()
    }

    fun onSeriesFollowClicked() {
        if (series == null) return

        if (series?.loggedInUserFollows != true) {
            disposables.add(seriesRepository.setFollowSeries(series!!)
                    .observeOn(schedulerProvider.ui())
                    .subscribe({ series: Series? -> onSeriesFollowed(series) }) { throwable: Throwable? -> onSeriesFollowedError(throwable) })
        } else {
            disposables.add(seriesRepository.unfollowSeries(series!!)
                    .observeOn(schedulerProvider.ui())
                    .subscribe({ series: Series? -> onSeriesFollowed(series) }) { throwable: Throwable? -> onSeriesFollowedError(throwable) })
        }
    }

    private fun onSeriesFollowedError(throwable: Throwable?) {
        view.get()?.showError("Error al dejar seguir la serie :(")
    }

    private fun onSeriesFollowed(series: Series?) {
        view.get()?.setSubtitle(series?.followers.toString() + if (series?.followers == 1) " follower" else " followers")
        view.get()?.setSeriesFollowed(series?.loggedInUserFollows)
    }

}